### ISSUE TITLE
[Enhancement] Air-gap operation: Support using a private registry for Rancher agent image

### ISSUE THREAD

**Issue by janeczku** [harvester/harvester#2176]
**Is your feature request related to a problem? Please describe.**
Harvester cluster is deployed in an air-gap network with a private container registry. When registering that cluster in Rancher, Harvester attempts (unsuccessfully) to download the `cattle-cluster-agent` image from Docker Hub.

Rancher allows specifying a "system-default-registry" to override, inter alia, the registry used to pull the agent images. But this is a global setting and applies to all Harvester clusters, which may not share a common registry.

**Expected behavior**

When registering a Harvester cluster in Rancher, the user may specify a private registry URL for pulling all required (agent) images.

**Comment by deskpil0t** [comment]
possible workaround.  download/migrate the image with zarf.  And assuming you have control over the CAs you can spoof the image/registry/domain-name for setup.  https://github.com/defenseunicorns/zarf

**Comment by guangbochen** [comment]
this issue seems duplicate with https://github.com/harvester/harvester/issues/2175? cc @FrankYang0529

**Comment by FrankYang0529** [comment]
I think it's different. #2175 is about CA cannot be applied to the host. This issue is about setting a private registry for Harvester cluster. For example, users may use a Rancher to manage multiple Harvester clusters, but the private registry for each Harvester cluster may be different.

**Comment by guangbochen** [comment]
Got it thanks, then we will need a setting to allow users to config a private registry and upload a custom CA cert of the Harvester cluster.

**Comment by Martin-Weiss** [comment]
We have to ensure that this feature allows full configuration of the Harvester integrated RKE2 registries.yaml with

1. registry mirror (multiple targets as contained supports registry fault tolerance / failover between registries provided)
3. registry namesapce mapping (i.e. image stored as registry01.customer.domain:5000/production/docker.io/rancher/rancher-agent:v2.6.6)
4. authentication (username and password)
5. and certificate authority bundle management (one global bundle i.e. /etc/ssl/ca-bundle.pem or separate ones per registry)

I assume this should be similar to the way how Rancher allows to configure this during RKE2 deployment:

![grafik](https://user-images.githubusercontent.com/13362573/176880833-e4943097-b96c-437b-87fe-420238cf1669.png)

**Comment by TachunLin** [comment]
Verifying **fixed** on `master-6e8e21b2-head` (10/04).

### Result
### Scenario 1
- [x] Update the Harvester containerd-registry setting with private registry, confirm **can** pull image and **deploy** nginx service correctly from private registry

  ![image](https://user-images.githubusercontent.com/29251855/193785720-970d9f18-76c6-4818-ae6a-e6239baf8ebe.png)
  ![image](https://user-images.githubusercontent.com/29251855/193787642-397ccf3a-144f-4785-9601-12a9017be325.png)
  ![image](https://user-images.githubusercontent.com/29251855/193787780-fc46a061-bbca-4051-839c-c4c092e67414.png)

### Scenario 2 
- [x] Change the Harvester `containerd-registry` to default value, the nginx 1.22 cannot be deployed since no private registry assigned. 

![image](https://user-images.githubusercontent.com/29251855/193790572-d5f5c92e-cbb9-4528-9c01-748d37940bc1.png)

### Scenario 3 

- [x] **Add back** the the Harvester containerd-registry setting to private registry, confirm **can** pull image and **deploy** nginx service correctly from private registry

  ![image](https://user-images.githubusercontent.com/29251855/193791385-5107385d-66c5-44a3-a298-82c44af929c7.png)

### Test Information
* Test Environment: 1 node Harvester on local kvm machine
* Harvester version: `master-6e8e21b2-head` (10/04)

### Verify Steps
#### Environment Setup 
Follow the environment preparation steps in https://github.com/harvester/tests/issues/476#issue-1355132277
* copy certs/domain.crt content in myregistry VM and paste it to additional-ca setting.
![image](https://user-images.githubusercontent.com/29251855/193785130-778991d2-6110-4dd5-ac63-e44f8201a30a.png)

#### Test plan 1
1. Update the Harvester containerd-registry setting to use private registry
  ![image](https://user-images.githubusercontent.com/29251855/193785720-970d9f18-76c6-4818-ae6a-e6239baf8ebe.png)
  ![image](https://user-images.githubusercontent.com/29251855/193785814-02c7b5b5-89da-43f0-a5f5-f36b8609a2df.png)
  ```
  {
  "Mirrors": {
    "docker.io": {
      "Endpoints": [
        "https://myregistry.local:5000"
      ],
      "Rewrites": null
    }
  },
  "Configs": {
    "myregistry.local:5000": {
      "Auth": null,
      "TLS": {
        "CAFile": "",
        "CertFile": "",
        "KeyFile": "",
        "InsecureSkipVerify": false
      }
    }
  },
  "Auths": null
}
  ```
2. Open K9s -> search -> secrets -> containerd -> y
  ![image](https://user-images.githubusercontent.com/29251855/193786543-ed43eb8a-f332-48b9-8b8a-afef8088f97f.png)
3. Check content in cattle-system/harvester-containerd-registry secret is changed. 
  ![image](https://user-images.githubusercontent.com/29251855/193787031-2ea11b68-e79c-4c10-8409-0bb337b5baf8.png)
4. Search jobs -> containerd -> l -> tail
5. Check there are new jobs to auto apply the new containerd-registry setting.
  ![image](https://user-images.githubusercontent.com/29251855/193786463-ca1e1f26-e37b-4e1b-b0d3-5e58b7d47bc1.png)
6. Apply following yaml in file 
    ```
    apiVersion: apps/v1
    kind: Deployment
    metadata:
      name: myregistry-nginx
    spec:
      selector:
        matchLabels:
          image: myregistry-nginx
      template:
        metadata:
          labels:
            image: myregistry-nginx
        spec:
          containers:
            - imagePullPolicy: Always
              image: myregistry.local:5000/nginx:latest
              name: nginx
    ```
    ```
    harvester-node-0:~ # kubectl apply -f deploy.yaml
    deployment.apps/myregistry-nginx created
    
    ```
7. search deployments -> default namespace -> myregsistry-nginx -> describe -> shift + g
8. Check the nginx can be deployed. 
  
#### Test plan 2 
1. Click the `default value` to update the Harvester containerd-registry setting to default empty 
  ![image](https://user-images.githubusercontent.com/29251855/193788085-0b1648d7-ef0e-495e-8898-4f743028f94c.png)
2. In K9s
3. Check content in cattle-system/harvester-containerd-registry secret is changed.
4. Check there are new jobs to auto apply the new containerd-registry setting.
5. Search secrets -> containerd -> e 
6. Clean registry.yaml data field
  ```
  # Please edit the object below. Lines beginning with a '#' will be ignored,
  # and an empty file will abort the edit. If an error occurs while saving this file will be
  # reopened with the relevant failures.
  #
  apiVersion: v1
  data:
    registries.yaml: ""
  kind: Secret
  metadata:
    creationTimestamp: "2022-10-04T08:51:47Z"
    name: harvester-containerd-registry
    namespace: cattle-system
    resourceVersion: "42088"
    uid: 8445547f-430a-474e-9783-dc6b16914eb7
  type: Opaque
  ```
7.  Search -> deployments -> find nginx-1.2.2
8.  Check cannot pull nginx-1.2.2 image
  ![image](https://user-images.githubusercontent.com/29251855/193790572-d5f5c92e-cbb9-4528-9c01-748d37940bc1.png)

#### Test plan 3 
1. Add back the Harvester containerd-registry setting to use private registry again 
  ![image](https://user-images.githubusercontent.com/29251855/193790937-a702730e-cb57-4f16-8950-76372a1859e3.png)
2. Check content in cattle-system/harvester-containerd-registry secret is changed.
  ![image](https://user-images.githubusercontent.com/29251855/193791120-755347db-46d7-427b-9e17-adb77025b7c8.png)
3. Check there are new jobs to auto apply the new containerd-registry setting.
  ![image](https://user-images.githubusercontent.com/29251855/193791283-40985e8a-f450-4766-9f0c-68e708c643cb.png)
4. Check nginx 1.22 deployment is automatically back to running.
  ![image](https://user-images.githubusercontent.com/29251855/193791385-5107385d-66c5-44a3-a298-82c44af929c7.png)

**Comment by TachunLin** [comment]
There is an issue need to further check while we are verifying test plan 2: 

1. When we already have setup valid setting in Harvester `containerd-registry`, after clicking the `Use default value` and save. 
2. We found the `status` field of the containerd registry setting also been cleanup to empty in K9s 
3. Which would cause no new `apply-sync-containerd` jobs trigger to update the containerd setting  

![image](https://user-images.githubusercontent.com/29251855/193995299-4969720a-d6dd-4b5b-a240-be182229d923.png)

Since the original test plan 2 can be achieved by cleaning the status in registry.yaml data field. 

I suggest to close this issue and track the new issue we found in https://github.com/harvester/harvester/issues/2873
